### PR TITLE
MSBuild: Details on Task parameters

### PR DESCRIPTION
**Pull request by rainersigwald**
Expounded on how the MSBuild engine actually calls into task classes.

cc @jaredpar, @rikkigibson, who asked about this offline

@Mikejo5000

### CODE REVIEW

**Comment by PRMerger15** [comment]
@rainersigwald : Thanks for your contribution! The author(s) have been notified to review your proposed change.

File: docs/msbuild/task-writing.md [hunk 1]
```diff
@@ -135,6 +135,30 @@ public string RequiredProperty { get; set; }
 
  The `[Required]` attribute is defined by <xref:Microsoft.Build.Framework.RequiredAttribute> in the <xref:Microsoft.Build.Framework> namespace.
 
+## How [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] invokes a task
+
+When invoking a task, [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] first instantiates the task class, then calls that object's property setters for task parameters that are set in the task element in the project file. If the task element does not specify a parameter, or if the expression specified in the element evaluates to an empty string, the property setter is not called.
```
Comment (jaredpar) [inline docs/msbuild/task-writing.md]: What is this `[!INCLUDE[vstecmsbuild](...)]` syntax?

File: docs/msbuild/task-writing.md [hunk 2]
```diff
@@ -135,6 +135,30 @@ public string RequiredProperty { get; set; }
...
+</Project>
+```
+
+only the setter for `Input3` is called.
+
+A task should not depend on any relative order of parameter-property setter invocation.
+
+### Task parameter types
+
+The [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] natively handles properties of type `string`, `bool`, `ITaskItem` and `ITaskItem[]`. If a task accepts a parameter of a different type, [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] will invoke <xref:System.Convert.ChangeType> to convert from `string` (with all property and item references expanded) to the destination type. If the conversion fails for any input parameter, [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] will emit an error and will not call the task's `Execute()` method.
```
Comment (Mikejo5000) [inline docs/msbuild/task-writing.md]:
```suggestion
The [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] natively handles properties of type `string`, `bool`, `ITaskItem` and `ITaskItem[]`. If a task accepts a parameter of a different type, [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] invokes <xref:System.Convert.ChangeType> to convert from `string` (with all property and item references expanded) to the destination type. If the conversion fails for any input parameter, [!INCLUDE[vstecmsbuild](../extensibility/internals/includes/vstecmsbuild_md.md)] emits an error and does not call the task's `Execute()` method.
```

**Comment by PRMerger9** [comment]
@rainersigwald : Thanks for your contribution! The author(s) have been notified to review your proposed change.

File: docs/msbuild/task-writing.md (hunk 1 above)
Comment (rainersigwald) [reply docs/msbuild/task-writing.md]: As far as I can tell it expands to . . . the localized marketing name for "MSBuild"?

**Comment by Mikejo5000** [comment]
#sign-off

File: docs/msbuild/task-writing.md (hunk 1 above)
Comment (jaredpar) [reply docs/msbuild/task-writing.md]: I never knew I wanted to know this but what is the localized marketing name for "MSBuild"? I'll accept answers for anything but English. 😉

**Pull request merged**